core/src/main/java/com/sap/ai/sdk/core/AiCoreService.java

```diff
@@ -0,0 +1,164 @@
+package com.sap.ai.sdk.core;
+
+import static com.sap.ai.sdk.core.DestinationResolver.AI_CLIENT_TYPE_KEY;
+import static com.sap.ai.sdk.core.DestinationResolver.AI_CLIENT_TYPE_VALUE;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.common.collect.Iterables;
+import com.sap.cloud.sdk.cloudplatform.connectivity.ApacheHttpClient5Accessor;
+import com.sap.cloud.sdk.cloudplatform.connectivity.DefaultHttpDestination;
+import com.sap.cloud.sdk.cloudplatform.connectivity.Destination;
+import com.sap.cloud.sdk.cloudplatform.connectivity.DestinationProperty;
+import com.sap.cloud.sdk.cloudplatform.connectivity.exception.DestinationAccessException;
+import com.sap.cloud.sdk.cloudplatform.connectivity.exception.DestinationNotFoundException;
+import com.sap.cloud.sdk.services.openapi.apiclient.ApiClient;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.client.BufferingClientHttpRequestFactory;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
+import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.web.client.RestTemplate;
+
+/** Connectivity convenience methods for AI Core. */
+@Slf4j
+@RequiredArgsConstructor
+public class AiCoreService implements AiCoreDestination {
+
+  final Function<AiCoreService, Destination> baseDestinationHandler;
+  final BiFunction<AiCoreService, Destination, ApiClient> clientHandler;
+  final BiFunction<AiCoreService, Destination, DefaultHttpDestination.Builder> builderHandler;
+
+  /** The default constructor. */
+  public AiCoreService() {
+    this(
+        AiCoreService::getBaseDestination,
+        AiCoreService::getApiClient,
+        AiCoreService::getDestinationBuilder);
+  }
+
+  @Nonnull
+  @Override
+  public ApiClient client() {
+    final var destination = destination();
+    return clientHandler.apply(this, destination);
+  }
+
+  @Nonnull
+  @Override
+  public Destination destination() {
+    final var dest = baseDestinationHandler.apply(this);
+    return builderHandler.apply(this, dest).build();
+  }
```

Comment:
Can we reduce code duplication between `AiCoreService` and `AiCoreDeployment`? (not by adding methods but changing the structure)

Reply:
API differences

| AiCoreService           | AiCoreDeployment  |
|-------------------------|-------------------|
| forDeployment           | withResourceGroup |
| forDeploymentByModel    |                   |
| forDeploymentbyScenario |                   |
| withDestination |                   |


API commons as represented by interface `AiCoreDestination`:

| AiCoreService           | AiCoreDeployment  |
|-------------------------|-------------------|
| client() | client() |
| destination() | destination()                  |


I don't know how I can reduce the code while still keeping the differences and similarities.

Reply:
I will create a follow-up PR for this